tools/TestGeneratedKeys.cpp

```diff
@@ -0,0 +1,967 @@
+  std::cout << "All threshold cryptography tests were successful.\n";
+
+  return true;
+}
+
+// Function to simplify the process of freeing the dynamically allocated memory
+// referenced by each ReplicaConfig struct, which the main function below may
+// need to do in one of several places because it may return early in the event
+// of a failure.
+static void freeConfigs(const std::vector<bftEngine::ReplicaConfig>& configs) {
+  for (auto config: configs) {
+    if (config.thresholdSignerForExecution) {
+      delete config.thresholdSignerForExecution;
+    }
+    if (config.thresholdVerifierForExecution) {
+      delete config.thresholdVerifierForExecution;
+    }
+    if (config.thresholdSignerForSlowPathCommit) {
+      delete config.thresholdSignerForSlowPathCommit;
+    }
+    if (config.thresholdVerifierForSlowPathCommit) {
+      delete config.thresholdVerifierForSlowPathCommit;
+    }
+    if (config.thresholdSignerForCommit) {
+      delete config.thresholdSignerForCommit;
+    }
+    if (config.thresholdVerifierForCommit) {
+      delete config.thresholdVerifierForCommit;
+    }
+    if (config.thresholdSignerForOptimisticCommit) {
+      delete config.thresholdSignerForOptimisticCommit;
+    }
+    if (config.thresholdVerifierForOptimisticCommit) {
+      delete config.thresholdVerifierForOptimisticCommit;
+    }
+  }
+}
+
+// Helper function for determining whether --help was given.
+static bool containsHelpOption(int argc, char** argv) {
+  for (int i = 1; i < argc; ++i) {
+    if (std::string(argv[i]) == "--help") {
+      return true;
+    }
+  }
+  return false;
+}
+
+/**
+ * Main function for the TestGeneratedKeys executable. TestGeneratedKeys is
+ * intended as a test for the correctness of GenerateConcordKeys, although it
+ * can also validate a set of keyfiles that have been manually created or
+ * modified as long as they comply with the output keyfile naming convention
+ * that GenerateConcordKeys uses.
+ *
+ * @param argc The number of arguments to this executable, including the name by
+ *             which this executable was launched as the first by convention.
+ * @param argv Command line arguments to this executable, including the name by
+ *             which it was launched as the first argument per convention. This
+ *             executable can be run with the --help option for usage
+ *             information on what command line arguments are expected.
+ *
+ * @return 0 if the testing is launched and completed successfully; -1 if the
+ *         tests could not be launched (for example, because of issue with the
+ *         command line parameters or missing input files) or if the input key
+ *         files fail the tests.
+ */
+int main(int argc, char** argv) {
+  std::string usageMessage = "Usage:\n"
+    "  TestGeneratedKeys -n TOTAL_NUMBER_OF_REPLICAS -o KEYFILE_PREFIX.\n"
+    "TestGeneratedKeys is intended to test the output of the"
+    " GenerateConcordKeys\nutility; TestGeneratedKeys expects that"
```

Comment:
I have tried reformatting this as you suggested; it will be reflected in the next round of revisions I push. Please tell me if you have any thoughts on the new formatting once it is pushed.